File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    static void func(List<String> result,String s,int open,int close,int n){
        if(open==n && close==n){
            result.add(s);
            return;
        }
        if(open<n){
            func(result,s+"(",open+1,close,n);
        }
        if(close<open){
            func(result,s+")",open,close+1,n);
        }
    }
    public List<String> generateParenthesis(int n) {
        List<String> result=new ArrayList<>();
        func(result,"",0,0,n);
        return result;
    }
}

File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
    void recur(int n, int open, int close, string s, vector<string>& ans) {
        if (s.size() == 2 * n && open == close) {
            ans.push_back(s);
            return;
        }
        if (open < n) {
            recur(n, open + 1, close, s + '(', ans);
        }

        if (open > close) {
            recur(n, open, close + 1, s + ')', ans);
        }
    }

    vector<string> generateParenthesis(int n) {
        int open = 0;
        int close = 0;
        vector<string> ans;
        string s = "";  // Initialize s as an empty string
        recur(n, open, close, s, ans);
        return ans;
    }
};